### PR TITLE
[3.3.5]Scripts/SunkenTemple: Quest 3447 Secret of the Circle

### LINKED ISSUE #808: Sunken Temple Instance (Put Working Forcefield & Working Secret of The Circle Quest & Spawn)
CORE revision number : af0d7e415df2+
Database: TDB_335.11.37
Patches: None

Inside the instance of sunken temple there must be a forcefield that goes down when you killed the six troll bosses. And, There is a Idol and a mini Boss who spawns when you use the 6 pilars in the correct order.

Changes C++:
Added the correct light effect when you touch all the statues in the correct order. "UseLastStatue()" and spawns the MiniBoss Atalarion  & The Idol
Added a control to remove the forcefield when the 6 troll bosses will dead.

Changes SQL:
Inserts the Forcefield 
Change the IAs of the troll bosses from EventAI to SmartAI and add a SET state to send the dead to the C++ Script,
Delete Atak'arion Mini Boss spawn (It will be spawned when use the statues in the correct order)
Change The type & Faction of the Altar, now yo can complete the 3446 Quest.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/259053-sunken-temple-instance-put-working-forcefield-working-secret-of-the-circle-quest-spawn?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by durotar**
Thanks to Tkrokli, Kilyana and Nayd
Updates #808

Signed-off-by: durotar dannytj2000@gmail.com

### CODE REVIEW

**Comment by durotar** [comment]
http://www.wowhead.com/object=148830/atalai-statue#screenshots:id=149091

File: sql/updates/world/2016_01_23_XX_world.sql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+-- Mini-boss Atal'alarion <Guardian of the Idol> and GameObject Idol of Hakkar
+SET @ATAL_ALARION := 8580;
+SET @ATALAI_IDOL := 148838;
+SET @GO_GUID := XXXXX;
+DELETE FROM `creature` WHERE `guid`=34521 AND `id`=@ATAL_ALARION; -- spawned by script
+
+DELETE FROM `gameobject` WHERE `guid`=@GO_GUID AND `id`=@ATALAI_IDOL; -- spawned but hidden until creature die
+INSERT INTO `gameobject` (`guid`,`id`,`map`,`zoneid`,`areaid`,`spawnMask`,`phaseMask`,`position_x`,`position_y`,`position_z`,`orientation`,`rotation0`,`rotation1`,`rotation2`,`rotation3`,`spawntimesecs`,`animprogress`,`state`,`VerifiedBuild`) VALUES
+(@GO_GUID,@ATALAI_IDOL,109,0,0,3,1,-476.2693,94.41199,-189.7297,1.588249,0,0,0,1,7200,255,1,12340);
```
Comment (Killyana) [inline sql/updates/world/2016_01_23_XX_world.sql]: spawntimesecs must be -1 to be initially despawned.

**Comment by durotar** [comment]
oops

**Comment by durotar** [comment]
How to test this fix:
.quest add 3445 (for alliance) or xxxx for horde
complete it in tanaris... then accept this quest and go to sunken temple with .tele command
enter the instances walk to the right downstairs (test with .gm off)
after clic all the 6 statues the mini-boss will be spawned.when he is dead, the idol will appear and the quest can be completed.

**Comment by tkrokli** [comment]
Nice, good job! :+1: Do you mind if I comment abount some unneeded spaces to can be removed in the .cpp file (codestyle) and maybe alignment of the GUIDs in the SQL (cosmetic) ? :-)

**Comment by durotar** [comment]
of course, thank you

File: sql/updates/world/2016_01_23_XX_world.sql [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/* Sunken Temple - Support for quest 3447: Secret of the Circle */
+SET @GO_GUID := 9999999;
+SET @ALTAR := 148836;
+SET @LIGHT := 148883;
+SET @ATAL_ALARION := 8580;
```
Comment (tkrokli) [inline sql/updates/world/2016_01_23_XX_world.sql]: SQL files usually do not need text alignment like in the .cpp files,
but it is common practice to align a "set @ variables" section:
```sql
SET @GO_GUID      := 9999999;
SET @ALTAR        :=  148836;
SET @LIGHT        :=  148883;
SET @ATAL_ALARION :=    8580;
SET @ATALAI_IDOL  :=  148838;
```
The order of these variables can be sorted either by variable name, order of appearance in the SQL script or by value, but it is not important and usually chosen by the author. The GUID value to be set by TC is usually first, just like you have set it here.

File: src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp [hunk 3]
```diff
@@ -35,7 +35,6 @@ enum Gameobject
     GO_ATALAI_STATUE4           = 148833,
     GO_ATALAI_STATUE5           = 148834,
     GO_ATALAI_STATUE6           = 148835,
-    GO_ATALAI_IDOL              = 148836,
     GO_ATALAI_LIGHT1            = 148883,
     GO_ATALAI_LIGHT2            = 148937
```
Comment (tkrokli) [inline src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp]: Line 40 in your .cpp file (line 41 in the old version) is an unneeded blank line. Please remove it.

File: sql/updates/world/2016_01_23_XX_world.sql (hunk 2 above)
Comment (MitchesD) [inline sql/updates/world/2016_01_23_XX_world.sql]: we don't need such a codestyle on SQL

File: src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp [hunk 4]
```diff
@@ -98,7 +96,6 @@ class instance_sunken_temple : public InstanceMapScript
                 case GO_ATALAI_STATUE4: GOAtalaiStatue4 = go->GetGUID();   break;
                 case GO_ATALAI_STATUE5: GOAtalaiStatue5 = go->GetGUID();   break;
                 case GO_ATALAI_STATUE6: GOAtalaiStatue6 = go->GetGUID();   break;
```
Comment (tkrokli) [inline src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp]: According to TC codestyle, `switch` / `case` statements should have `break;` statements on their own line beneath the case line and with a 4 space indent. Not sure if it is preferred to leave it unchanged or make the improvement.

**Comment by durotar** [comment]
I chose to change as little as possible the code for better understanding of my patch. But it can be done in a general review of the files in this directory.

**Comment by tkrokli** [comment]
Good idea. Better to keep it simple and to the point. :+1:

**Comment by Rushor** [comment]
problems are resolved, any more feedback on this?

File: src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp [hunk 5]
```diff
@@ -169,18 +169,16 @@ class instance_sunken_temple : public InstanceMapScript
...
-             AtalaiStatue5->SummonGameObject(GO_ATALAI_LIGHT2, AtalaiStatue5->GetPositionX(), AtalaiStatue5->GetPositionY(), AtalaiStatue5->GetPositionZ(), 0, 0, 0, 0, 0, 100);
-             AtalaiStatue6->SummonGameObject(GO_ATALAI_LIGHT2, AtalaiStatue6->GetPositionX(), AtalaiStatue6->GetPositionY(), AtalaiStatue6->GetPositionZ(), 0, 0, 0, 0, 0, 100);
-             go->SummonGameObject(148838, -488.997, 96.61, -189.019, -1.52, 0, 0, 0, 0, 100);
+             go->SummonGameObject(GO_ATALAI_LIGHT2, -515.553f, 95.25821f, -173.707f, 0, 0, 0, 0, 0, 0);
+             go->SummonGameObject(GO_ATALAI_LIGHT2, -419.8487f, 94.48368f, -173.707f, 0, 0, 0, 0, 0, 0);
+             go->SummonGameObject(GO_ATALAI_LIGHT2, -491.4003f, 135.9698f, -173.707f, 0, 0, 0, 0, 0, 0);
+             go->SummonGameObject(GO_ATALAI_LIGHT2, -491.4909f, 53.48179f, -173.707f, 0, 0, 0, 0, 0, 0);
+             go->SummonGameObject(GO_ATALAI_LIGHT2, -443.8549f, 136.1007f, -173.707f, 0, 0, 0, 0, 0, 0);
+             go->SummonGameObject(GO_ATALAI_LIGHT2, -443.4171f, 53.83124f, -173.707f, 0, 0, 0, 0, 0, 0);
```
Comment (Treeston) [inline src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp]: Magic numbers are bad.

Comment (tkrokli) [inline src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp]: Can the float values be made into enums? ... or did you mean the zeros?

Comment (Kittnz) [inline src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp]: no.. just use Position

Comment (Treeston) [inline src/server/scripts/EasternKingdoms/SunkenTemple/instance_sunken_temple.cpp]:
````c++
static const Position lightPos[6] = {
    { -515.553f, 95.25821f, -173.707f },
    /* ... */
};
```

**Comment by MitchesD** [comment]
https://github.com/TrinityCore/TrinityCore/commit/768a5cf354b447e19e7e5b12c78dc4920d8b1caa

**Comment by durotar** [comment]
thank you very much

**Comment by tkrokli** [comment]
Looking forward to testing this result. :+1:

**Pull request closed without merging**